### PR TITLE
(feat) O3-2117: Save orders on a new encounter if there is no encounter

### PR DESCRIPTION
**Pull request by usamaidrsk**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing.en-US#contributing-guidelines) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://om.rs/o3ui).
- [ ] My work includes tests or is validated by existing tests.

## Summary

This PR adds the functionality to save orders along with creating a new encounter if one didn't exist. Before two transactions existed, i.e one to create encounter first and then create orders.

## Screenshots

https://github.com/openmrs/openmrs-esm-patient-chart/assets/53287480/a68fbd48-a32b-4e0d-977c-94b7e53161a1

## Related Issue

[O3-2117](https://openmrs.atlassian.net/browse/O3-2117)

## Other

[O3-2117]: https://openmrs.atlassian.net/browse/O3-2117?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/esm-patient-orders-app/src/api/api.ts
Comment (raghavawasthi24) [inline packages/esm-patient-orders-app/src/api/api.ts:89]: Instead of this `for loop`, you can use built-in react method `map`. Although functionality of both are almost same but `map` is faster than `for loop`

Comment (vasharma05) [reply packages/esm-patient-orders-app/src/api/api.ts:89]: Hi @raghavawasthi24!
I believe you mean to say `forEach`, since the `map` function is not a loop, but is used to map the item in an array to a custom value. `forEach` should be used if you don't want to return anything from the loop.

File: packages/esm-patient-orders-app/src/order-basket/order-basket.workspace.tsx
Comment (denniskigen) [inline packages/esm-patient-orders-app/src/order-basket/order-basket.workspace.tsx:130]: [InlineNotification](https://react.carbondesignsystem.com/?path=/docs/components-notifications-inline--overview) does not accept a prop called `inline`.

File: packages/esm-patient-orders-app/src/order-basket/order-basket.workspace.tsx
Comment (denniskigen) [inline packages/esm-patient-orders-app/src/order-basket/order-basket.workspace.tsx:119]: I think it's best to show the original message here instead of the error message from the backend, which we're already logging to the console on line 67.

File: packages/esm-patient-orders-app/src/order-basket/order-basket.workspace.tsx
Comment (denniskigen) [inline packages/esm-patient-orders-app/src/order-basket/order-basket.workspace.tsx:126]: Nit:

```suggestion
              title={t('errorCreatingAnEncounter', 'Error creating an encounter')}
```

File: packages/esm-patient-orders-app/src/order-basket/order-basket.workspace.tsx
Comment (denniskigen) [inline packages/esm-patient-orders-app/src/order-basket/order-basket.workspace.tsx:81]: Is it preferable to await the promise instead?

```suggestion
      await mutateOrders()
```

File: packages/esm-patient-orders-app/src/order-basket/order-basket.workspace.tsx
Comment (denniskigen) [inline packages/esm-patient-orders-app/src/order-basket/order-basket.workspace.tsx:36]: `isSavingOrders` reads clearer to me here:

```suggestion
  const [isSavingOrders, setIsSavingOrders] = useState(false);
```

**Review by denniskigen — COMMENTED** [review]
Thanks for working on this, @usamaidrsk. I've left a few suggestions

File: packages/esm-patient-orders-app/src/order-basket/order-basket.workspace.tsx
Comment (usamaidrsk) [reply packages/esm-patient-orders-app/src/order-basket/order-basket.workspace.tsx:119]: I think that's right. 
Just what I thought that given on creating the encounter we're  adding orders too, then if the error was from the orders, showing the exact message might be helpful to the user what exactly is the issue.

File: packages/esm-patient-orders-app/src/api/api.ts
Comment (vasharma05) [inline packages/esm-patient-orders-app/src/api/api.ts:88]: Instead of getting the `orders` from the store, the orders should be passed via the function arguments.

Comment (vasharma05) [reply packages/esm-patient-orders-app/src/api/api.ts:88]: Also, the function `createEmptyEncounter` should be renamed to `saveOrdersWithNewEncounter`, since we are also saving the orders in the same encounter request.

**Review by vasharma05 — CHANGES_REQUESTED** [review]
A few changes

File: packages/esm-patient-orders-app/src/api/api.ts
Comment (samuelmale) [inline packages/esm-patient-orders-app/src/api/api.ts:12]: Is it possible to use a single import statement?

**Review by samuelmale — APPROVED** [review]
LGTM

File: packages/esm-patient-orders-app/src/api/api.ts
Comment (denniskigen) [reply packages/esm-patient-orders-app/src/api/api.ts:12]: You'll likely need to make sure that those exports are re-exported from https://github.com/openmrs/openmrs-esm-patient-chart/blob/main/packages/esm-patient-common-lib/src/orders/index.ts first. That might need to happen in a separate PR.

File: packages/esm-patient-orders-app/src/api/api.ts
Comment (ibacher) [reply packages/esm-patient-orders-app/src/api/api.ts:82]: You can even save a line:

```ts
  Object.entries(patientItems).forEach(([grouping, groupOrders]) => {
    groupOrders.forEach(order => {
      orders.push(postDataPrepFunctions[grouping](order, patientUuid, null));
    });
  });
```

File: packages/esm-patient-orders-app/src/api/api.ts
Comment (usamaidrsk) [reply packages/esm-patient-orders-app/src/api/api.ts]: As discussed with @brandones , moving the implementation of saving the encounter and orders to the `common-lib` seemed like a better idea. Therefore, I have created the `postOrdersOnNewEncounter` function in the [postOrders.ts](https://github.com/usamaidrsk/openmrs-esm-patient-chart/blob/f5e5969471386920761f3b1dbae443c377d2a297/packages/esm-patient-common-lib/src/orders/postOrders.ts#L6) file.

cc @ibacher @denniskigen @samuelmale

File: packages/esm-patient-common-lib/src/orders/postOrders.ts
Comment (brandones) [inline packages/esm-patient-common-lib/src/orders/postOrders.ts:21]:
```suggestion
      'postOrdersOnNewEncounter received an active visit that is not currently active. This is a programming error. Attempting to place the order using the visit start date.',
```

File: packages/esm-patient-orders-app/src/order-basket/order-basket.workspace.tsx
Comment (brandones) [reply packages/esm-patient-orders-app/src/order-basket/order-basket.workspace.tsx:120]: Dennis may be the leading authority on UI copy; I would defer to him @usamaidrsk .

**Comment by brandones** [comment]
Please also modify [here](https://github.com/openmrs/openmrs-esm-patient-chart/blob/8f04eba09c0f17658c39ed01dafae087e4422c9c/packages/esm-patient-medications-app/src/api/api.ts#L59) and [here](https://github.com/openmrs/openmrs-esm-patient-chart/blob/main/packages/esm-patient-labs-app/src/lab-orders/api.ts#L78) with the optional encounterUuid type.

File: packages/esm-patient-orders-app/src/api/api.ts
Comment (brandones) [inline packages/esm-patient-orders-app/src/api/api.ts:44]: This function is duplicated in common-lib; let's delete this one.

**Review by brandones — CHANGES_REQUESTED** [review]

**Review by brandones — APPROVED** [review]
LGTM

**Comment by brandones** [comment]
@jayasanka-sack The E2E tests kept getting canceled? Any idea what that's about? Can we go ahead and merge this or would we risk breaking the E2E tests?

**Comment by jayasanka-sack** [comment]
Hey, I checked out the recordings, and it looks like the UI isn't doing anything after clicking the save button. Since other PRs and the main branch don't have this issue, it seems like this problem was introduced in this PR.

Recordings:

[lab order](https://github.com/openmrs/openmrs-esm-patient-chart/assets/33048395/6c13ce93-0ae4-4684-aba2-8b1ee4ed38f5)
[drug order](https://github.com/openmrs/openmrs-esm-patient-chart/assets/33048395/0b93a8bc-f0cf-4c7d-8b35-861df43a3705)

**Comment by brandones** [comment]
Thanks @jayasanka-sack ! Ok @usamaidrsk we need to fix the E2E tests.

**Comment by jayasanka-sack** [comment]
I ran the test locally but couldn't reproduce this behavior. The key difference when running E2E tests on GitHub compared to running them locally is that we don't use the dev server on GitHub. Instead, we build the frontend and serve it, which is somewhat similar to the dev3 environment.

However, I reviewed the [trace](https://playwright.dev/docs/ci-intro#viewing-the-trace) from the report and found these console errors:

```
TypeError: (0 , c.postOrdersOnNewEncounter) is not a function
    at http://localhost:8080/openmrs/spa/openmrs-esm-patient-orders-app-8.0.1/287.js:2:14249
    at http://localhost:8080/openmrs/spa/openmrs-esm-patient-orders-app-8.0.1/287.js:2:13958
    at Object.next (http://localhost:8080/openmrs/spa/openmrs-esm-patient-orders-app-8.0.1/287.js:2:14063)
    at S (http://localhost:8080/openmrs/spa/openmrs-esm-patient-orders-app-8.0.1/287.js:2:9123)
    at a (http://localhost:8080/openmrs/spa/openmrs-esm-patient-orders-app-8.0.1/287.js:2:9326)
    at http://localhost:8080/openmrs/spa/openmrs-esm-patient-orders-app-8.0.1/287.js:2:9385
    at new Promise (<anonymous>)
    at http://localhost:8080/openmrs/spa/openmrs-esm-patient-orders-app-8.0.1/287.js:2:9266
    at Object.Re (http://localhost:8080/openmrs/spa/openmrs.167c04e490dfd1aa.js:2:954551)
    at Ve (http://localhost:8080/openmrs/spa/openmrs.167c04e490dfd1aa.js:2:954705)
Unhandled rejection:  TypeError: Cannot read properties of undefined (reading 'error')
    at http://localhost:8080/openmrs/spa/openmrs-esm-patient-orders-app-8.0.1/287.js:2:14503
    at http://localhost:8080/openmrs/spa/openmrs-esm-patient-orders-app-8.0.1/287.js:2:13958
    at Object.next (http://localhost:8080/openmrs/spa/openmrs-esm-patient-orders-app-8.0.1/287.js:2:14063)
    at S (http://localhost:8080/openmrs/spa/openmrs-esm-patient-orders-app-8.0.1/287.js:2:9123)
    at a (http://localhost:8080/openmrs/spa/openmrs-esm-patient-orders-app-8.0.1/287.js:2:9326)
    at http://localhost:8080/openmrs/spa/openmrs-esm-patient-orders-app-8.0.1/287.js:2:9385
    at new Promise (<anonymous>)
    at http://localhost:8080/openmrs/spa/openmrs-esm-patient-orders-app-8.0.1/287.js:2:9266
    at Object.Re (http://localhost:8080/openmrs/spa/openmrs.167c04e490dfd1aa.js:2:954551)
    at Ve (http://localhost:8080/openmrs/spa/openmrs.167c04e490dfd1aa.js:2:954705)
showToast must be called with an object having a 'description' property that is a non-empty string or object
Cannot read properties of undefined (reading 'error')
TypeError: Cannot read properties of undefined (reading 'error')
    at http://localhost:8080/openmrs/spa/openmrs-esm-patient-orders-app-8.0.1/287.js:2:14503
    at http://localhost:8080/openmrs/spa/openmrs-esm-patient-orders-app-8.0.1/287.js:2:13958
    at Object.next (http://localhost:8080/openmrs/spa/openmrs-esm-patient-orders-app-8.0.1/287.js:2:14063)
    at S (http://localhost:8080/openmrs/spa/openmrs-esm-patient-orders-app-8.0.1/287.js:2:9123)
    at a (http://localhost:8080/openmrs/spa/openmrs-esm-patient-orders-app-8.0.1/287.js:2:9326)
    at http://localhost:8080/openmrs/spa/openmrs-esm-patient-orders-app-8.0.1/287.js:2:9385
    at new Promise (<anonymous>)
    at http://localhost:8080/openmrs/spa/openmrs-esm-patient-orders-app-8.0.1/287.js:2:9266
    at Object.Re (http://localhost:8080/openmrs/spa/openmrs.167c04e490dfd1aa.js:2:954551)
    at Ve (http://localhost:8080/openmrs/spa/openmrs.167c04e490dfd1aa.js:2:954705)
```

**Comment by usamaidrsk** [comment]
> Meanwhile, @usamaidrsk I wonder why the UI silenced the error. We should display an error message to the user if something goes wrong.

Thanks for the follow up @jayasanka-sack , however all errors are displayed by an inline notification on the UI. Not sure why it is skipping that on the tests

File: packages/esm-patient-orders-app/src/order-basket/order-basket.workspace.tsx
Comment (jayasanka-sack) [inline packages/esm-patient-orders-app/src/order-basket/order-basket.workspace.tsx:76]: > Thanks for the follow up @jayasanka-sack , however all errors are displayed by an inline notification on the UI. Not sure why it is skipping that on the tests

@usamaidrsk I don't think it has anything to do with the test. The issue lies on this line I guess. In this specific instance theres no `error` property on the `e.responseBody`.

You can find it's staten on the error log:
```
Unhandled rejection:  TypeError: Cannot read properties of undefined (reading 'error')

```
```
showToast must be called with an object having a 'description' property that is a non-empty string or object
Cannot read properties of undefined (reading 'error')
```

**Comment by brandones** [comment]
Ok, since this is an E2E bug I'm going to go ahead and merge.

**Comment by denniskigen** [comment]
Would you expect E2E tests to pass once this got merged in, @brandones? Still failing for new PRs surprisingly.

**Comment by ibacher** [comment]
They should, but unfortunately the patient-common-lib is embedded in lots of places that are not the chart, so we may not be using the latest version in tests.

**Pull request merged**
